### ISSUE TITLE
websockets next - Multi-Tenant / RoutingContext

### ISSUE THREAD

**Issue by Paul6552** [quarkusio/quarkus#42737]
### Describe the bug

I tried the new websockets next with panache but it throws several exceptions. First when I try to evaluation the value from this Code line: `Setting test = Setting.findAll().firstResult();`
then I get the following error message:
`This method is normally automatically overridden in subclasses: did you forget to annotate your entity with @Entity?`

When I leave the program running I enter the exception and I get the following error message:
`Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()`

### Expected behavior

No error, returning a int-number.

### Actual behavior

Error message

### How to Reproduce?

I have this websocket next endpoint:

```
@WebSocket(path = "/beamer/{uselessID}", endpointId = "Beamer")
@ApplicationScoped
public class BeamerWebsocketEndpoint {

    @OnOpen()
    public Integer onOpen() {
        try {
            Setting test = Setting.findAll().firstResult(); //Stop the debugger here and evaluate the value Setting.findAll....
            return test.getRealNumber();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
        }
        return 222;
    }

    @OnClose
    public void onClose() {
        System.out.println("Someone left the game");
    }

    @OnError
    String error(Exception t) {
        System.err.println("Error: " + t.getMessage());
   
        return "error";
    }
}
```

This is my `Setting.java` class:
```
@Entity
@NoArgsConstructor
@Data
@EqualsAndHashCode(callSuper=false)
public class Setting extends PanacheEntityBase implements Serializable {

    @Id
    @GeneratedValue( strategy = GenerationType.AUTO )
    private long einstellung_id;

    @Basic
    @Column(name = "realNumber")
    private int realNumber;
}
```

I try the connection with the quarkus developer ui:
http://localhost:8080/q/dev-ui/io.quarkus.quarkus-websockets-next/server-endpoints

Connect to the websocket next endpoint and you get the two error messages explained above.

Behind is a postgresql database and in the setting-table there is a row in the database.
With a normal rest-response call `Setting test = Setting.findAll().firstResult();` returns a number.

If needed my pom.xml file:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <artifactId>pea-backend</artifactId>

    <parent>
        <groupId>at.tanzauswertung</groupId>
        <artifactId>pea</artifactId>
        <version>2.0.0-SNAPSHOT</version>
    </parent>

    <properties>
        <maven.compiler.source>21</maven.compiler.source>
        <maven.compiler.target>21</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <!-- Update quarkus
        https://quarkus.io/guides/update-quarkus
        Dort nimm das "Using maven" Befehl und für diesen im CMD unter pea-backend aus -->
        <quarkus.version>3.13.2</quarkus.version>
        <version.maven.surefire-plugin>3.2.3</version.maven.surefire-plugin>
        <version.maven.failsafe-plugin>3.1.0</version.maven.failsafe-plugin>
        <version.apache-poi>5.2.5</version.apache-poi>
    </properties>

    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>io.quarkus.platform</groupId>
                <artifactId>quarkus-bom</artifactId>
                <version>${quarkus.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>

    <dependencies>
        <dependency>
            <groupId>at.tanzauswertung.api</groupId>
            <artifactId>pea-portal-api</artifactId>
            <version>1.1-SNAPSHOT</version>
        </dependency>
        <dependency>
            <groupId>org.apache.poi</groupId>
            <artifactId>poi</artifactId>
            <version>${version.apache-poi}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.poi</groupId>
            <artifactId>poi-ooxml</artifactId>
            <version>${version.apache-poi}</version>
        </dependency>
        <dependency>
            <groupId>org.flywaydb</groupId>
            <artifactId>flyway-database-postgresql</artifactId>
        </dependency>
        <dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <version>1.18.32</version>
            <scope>provided</scope>
        </dependency>
        <dependency>
            <groupId>net.sf.jasperreports</groupId>
            <artifactId>jasperreports</artifactId>
            <version>6.21.3</version>
            <type>jar</type>
            <exclusions>
                <exclusion>
                    <artifactId>javax.inject</artifactId>
                    <groupId>javax.inject</groupId>
                </exclusion>
            </exclusions>
        </dependency>
        <!-- glaube braucht man nicht mehr
        <dependency>
            <groupId>org.jboss.resteasy</groupId>
            <artifactId>resteasy-multipart-provider</artifactId>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.datatype</groupId>
            <artifactId>jackson-datatype-jsr310</artifactId>
            <version>2.14.2</version>
        </dependency>-->
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-narayana-jta</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-websockets-next</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-arc</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-scheduler</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-rest-client</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-rest-client-jackson</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-rest</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-rest-jackson</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-hibernate-orm</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-hibernate-orm-panache</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-flyway</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-jdbc-postgresql</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-agroal</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-oidc</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-undertow</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-keycloak-authorization</artifactId>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-junit5</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-test-h2</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.assertj</groupId>
            <artifactId>assertj-core</artifactId>
            <version>3.25.3</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>io.rest-assured</groupId>
            <artifactId>rest-assured</artifactId>
            <scope>test</scope>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-install-plugin</artifactId>
                <version>3.1.1</version>
            </plugin>
            <plugin>
                <groupId>io.quarkus</groupId>
                <artifactId>quarkus-maven-plugin</artifactId>
                <version>${quarkus.version}</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>build</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-resources-plugin</artifactId>
                <version>3.3.1</version>
                <executions>
                    <execution>
                        <id>copy-resources</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>copy-resources</goal>
                        </goals>
                        <configuration>
                            <outputDirectory>${project.build.directory}/classes/META-INF/resources/pea</outputDirectory>
                            <resources>
                                <resource>
                                    <directory>${project.parent.basedir}/pea-frontend/build/</directory>
                                </resource>
                            </resources>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-failsafe-plugin</artifactId>
                <version>${version.maven.failsafe-plugin}</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>integration-test</goal>
                            <goal>verify</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>${version.maven.surefire-plugin}</version>
                <configuration>
                    <systemPropertyVariables>
                        <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
                    </systemPropertyVariables>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.12.1</version>
                <configuration>
                    <compilerArgs>
                        <arg>-Xpkginfo:always</arg>
                    </compilerArgs>
                </configuration>
            </plugin>
        </plugins>
    </build>

</project>

```

### Output of `java -version`

21.0.3

### Quarkus version or git rev

3.13.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvn version 3.9.8

**Comment by mkouba** [comment]
@Paul6552 I was not able to reproduce the problem with a modified [websockets-next-quickstart](https://github.com/quarkusio/quarkus-quickstarts/tree/main/websockets-next-quickstart) - just added `quarkus-hibernate-orm-panache`, `quarkus-jdbc-postgresql` and `quarkus-narayana-jta` extensions, the `Setting` entity, and tried to call `Setting.findAll().firstResult()` in the `@OnOpen` callback.

I'd try to remove the `org.projectlombok:lombok` dependency which is usually troublesome in Quarkus apps and also not needed for a Panache entity.

In any case, it does not seem to be a problem of `quarkus-websockets-next`.

By the way, is the `quarkus-undertow` extension needed in your app? It should not be unless you need to work with Servlet API.

**Comment by mkouba** [comment]
CC @FroMage Does the error message `This method is normally automatically overridden in subclasses: did you forget to annotate your entity with @Entity?` in combination with lombok ring a bell?

**Comment by Paul6552** [comment]
@mkouba thanks Martin for looking into this. Today my holiday starts therefore I will reply in two weeks again. 

I will try to remove Lombok from my project, maybe that is the problem.

When nothing works I will provide you my code so maybe you find something in the original code. Thanks a lot

**Comment by FroMage** [comment]
I'm pretty sure Lombok does something weird in this case, but not sure what.

**Comment by Paul6552** [comment]
@mkouba @FroMage 

_I'd try to remove the org.projectlombok:lombok dependency which is usually troublesome in Quarkus apps and also not needed for a Panache entity._
I'm a little confused because we create a lot of projects with Quarkus + Panache + Lombok. We didn't see anywhere in the documentation that this is not supported or recommended! It would be **very important** to document this somewhere if this is really the case!

Unfortunately, we couldn't completely remove Lombok as a dependency because we need some functionality for DTOs (e.g. getter, setter, builder). Does Quarkus offer an alternative without writing all the getters/setters/builders yourself?

At least we were able to rewrite all class entities (@Entity) so that these classes ONLY use Panache.
We then tested websockets-next again with the same error.

**BUT we were able to identify the problem and it was our Schema/TenantID routing.**
Why we didn't think of it earlier, however, we saw a big difference to other apps: **schemas / TenantIDs**.
We turned off our routing and it worked!

Now the problem is/was that `websockets-next` apparently can't handle routing.

We have the following code:
```
@RequestScoped
@PersistenceUnitExtension
public class CustomTenantResolver implements TenantResolver {

    @Inject
    RoutingContext context;

    @Override
    public String getDefaultTenantId() {
        return "public";
    }

    @Override
    public String resolveTenantId() {
        String schema = context.request().getHeader( "x-schema" );

        if ( schema == null || schema.equalsIgnoreCase( "public" ) ) {
            return "public";
        }

        return "s" + schema;
    }
}
```

If we swap the code to this:
```
@RequestScoped
@PersistenceUnitExtension
public class CustomTenantResolver implements TenantResolver {

    @Override
    public String getDefaultTenantId() {
        return "public";
    }

    @Override
    public String resolveTenantId() {
        return "public";
    }
}
```

**everything works!!**

Now can you please explain to me/us how we can use `websockets-next` with multiple schemes? The client currently uses the header to indicate which schema it is.

**Comment by FroMage** [comment]
OK, so apparently it's not lombok, sorry about that. I didn't know you were using Hibernate multi-tenancy.

What options are there to obtain the headers with websockets, @mkouba @cescoffier ?

**Comment by mkouba** [comment]
Yes, so the problem is that websockets-next does not set the current Vert.x routing context via `io.quarkus.vertx.http.runtime.CurrentVertxRequest.setCurrent(RoutingContext)`. And this `RoutingContext` is injected in the `TenantResolver`.

> What options are there to obtain the headers with websockets, @mkouba @cescoffier ?

Within an endpoint callback invocation you can inject the initial handshake request with `@Inject io.quarkus.websockets.next.HandshakeRequest` and access the headers with `HandshakeRequest#header()` method.

**Comment by cescoffier** [comment]
It will have to be the initial routing context, but I'm a bit worried. That routing context is almost useless (or should not be used, actually). Why does it need it?

**Comment by mkouba** [comment]
> It will have to be the initial routing context, but I'm a bit worried. That routing context is almost useless (or should not be used, actually). Why does it need it?

FTR the initial `RoutingContext` is "there" but can't be injected...

**Comment by FroMage** [comment]
> Why does it need it?

For access to headers.

**Comment by FroMage** [comment]
I'm assuming here that the headers determine the DB tenant for a websocket, and can't change during the lifetime of the websocket.

**Comment by Paul6552** [comment]
> I'm assuming here that the headers determine the DB tenant for a websocket, and can't change during the lifetime of the websocket.

If that was a question directed at me, then I can confirm it. The web browser (client) uses the header to provide information about which database schema it wants to access. The schema cannot and should not be changed during the lifespan of this web socket.

Something to the comment from @mkouba  to` inject WebSocketConnection` and access the header via 
`connection.handshakeRequest().header();`

I haven't tested it yet but then there must be a way to cancel the function in the `TenantResolver` if it is a web socket. What I mean is that normal rest endpoints that need to access the tenant resolver should use the `RoutingContext`, but websockets can omit it. In terms of programming, what I mean is written in layman's terms:

```
@RequestScoped
@PersistenceUnitExtension
public class CustomTenantResolver implements TenantResolver {

    @Inject
    RoutingContext context;

    @Override
    public String getDefaultTenantId() {
        return "public";
    }

    @Override
    public String resolveTenantId() {
        if( **IF THIS IS A NORMAL REST ENDPOINT CALL USE ROUTINGCONTEXT**){
        String schema = context.request().getHeader( "x-schema" );

        if ( schema == null || schema.equalsIgnoreCase( "public" ) ) {
            return "public";
        }

        return "s" + schema;
       }else{
        // IT IS A WEBSOCKET - NOTHING TO DO. HEADERS will be checked in the websockets-next Endpoint with
        // connection.handshakeRequest().header();
       }
}
```

**Comment by FroMage** [comment]
You can probably use injection programatically in `resolveTenantId()` and try to look up `RoutingContext` and use it, which means you're in a regular HTTP call, or look up `WebSocketConnection` and use it if you're in a web-socket.

**Comment by Paul6552** [comment]
But how can I look up for RoutingContext !!

I tried a lot, but I get always the error message: 
`Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()`

```
@RequestScoped
@PersistenceUnitExtension
public class CustomTenantResolver implements TenantResolver {

    @Inject
    RoutingContext context;
    @Inject
    WebSocketConnection connection;

    @Override
    public String getDefaultTenantId() {
        return "public";
    }

    @Override
    public String resolveTenantId() {
        //Wir müssen hier zwischen einem normal Rest-Endpoint call und Websockets unterscheiden
        String schema;
        if(context != null && context.request() != null && context.session() != null && !context.failed() && context.session() != null) {
            schema = context.request().getHeader( "x-schema" );
        }else{
            //Ist ein Websocket
            schema = connection.handshakeRequest().header("x-schema");
        }

        if ( schema == null || schema.equalsIgnoreCase( "public" ) ) {
            return "public";
        }

        return "s" + schema;
    }
}
```
What do I have to check if `RoutingContext` exits or not?

**Comment by Paul6552** [comment]
Is this the correct code to look up if RoutingContext is valid or not? It looks wrong, but the goog news is it is working!
 ```
   @Override
    public String resolveTenantId() {
        //Wir müssen hier zwischen einem normal Rest-Endpoint call und Websockets unterscheiden
        String schema;
        try {
            schema = context.request().getHeader( "x-schema" );
        } catch (Exception e) {
            // Handle WebSocket logic
            schema = "106985"; //connection.handshakeRequest().header("x-schema");
        }

        if ( schema == null || schema.equalsIgnoreCase( "public" ) ) {
            return "public";
        }

        return "s" + schema;
    }
```

I think this issue can be closed, but for others it would be great to document this on the websocket-next reference guide:
https://quarkus.io/guides/websockets-next-reference

Thank you all

**Comment by mkouba** [comment]
> Is this the correct code to look up if RoutingContext is valid or not? It looks wrong, but the goog news is it is working!

Yes, it's not exactly correct. However, I'm not quite there's an easy way to create a `TenantResolver` that would work for both REST and WS next. The problem is that the request context is active (although it's a different context than the one used for the HTTP upgrade) and the `RoutingContext` bean is registered but "uninitialized".

In any case, we don't want to make the `RoutingContext` available in an `@OnOpen` callback because it's called after the HTTP upgrade is perfomed.

You can actually reverse the logic; i.e. `@Inject WebSocketConnection` and when `connection.handshakeRequest().header("x-schema")` throws `ContextNotActiveException` you can be sure that it's called outside a WS endpoint callback and therefore `context.request().getHeader( "x-schema" )` should work.

**Comment by Paul6552** [comment]
Maybe you can create more examples for such use cases or examples how to work woth uni, multi, schemas ...

There would also be an issue already:
https://github.com/quarkusio/quarkus/issues/42566

But I think there are only a few people (you?) which have knowledge in this new area websockets-next

Thanks for help

**Comment by FroMage** [comment]
Definitely worth mentioning in both Hibernate ORM's tenant docs (https://quarkus.io/guides/hibernate-orm#writing-the-application) and websockets-next. Would you be up for a docs PR @Paul6552 ?

**Comment by Paul6552** [comment]
I am an amateur programmer, so you are probably not happy with my PR because I don't always understand everything regarding RoutingContext and Websockets context but I wrote an article at websockets-next-reference.
https://github.com/quarkusio/quarkus/pull/43144

**Comment by FroMage** [comment]
Well, this is a great start, thanks a lot @Paul6552 I'm sure it will help others!
